### ISSUE TITLE
what are we doing?

### ISSUE THREAD

**Issue by sckott** [sckott/resteasy#1]
hey @cboettig  That idea you brought up the other day about having an API based on static files, without having to deal with servers. 

What about piggybacking on the [Github Contents API](https://developer.github.com/v3/repos/contents/)? So users can simply deal with csv files, anyone can do it, and an R client can do all the CRUD operations, so users can modify from R, get files, etc. 

Github API doesn't allow filtering or sorting, etc. by fields in a file, but one idea is to route a file through another API that allows this, e.g., https://github.com/project-open-data/csv-to-api, which I use in the function `easy_get()`. 

Anyway, just playing here, and easy to throw away, but was trying this approach to get close to the idea you had - this would be free, would be easy for small files, but even with large DB's could split the table(s) across many files - we could hide some tricks under the hood in the case of a single database split across many files.

**Comment by cboettig** [comment]
Hey @sckott ,

Nice, was just taking a look here, very cool.  

What do you see as the advantage of querying the Contents API vs just hosting the csv files via Github-pages? Seems like the latter would avoid the arbitrary limits of the API (1 MB files, 1000 files/dir, API call limits).  If the files exceed the 100 MB file limit or collectively the 1 GB repo limit (https://help.github.com/articles/working-with-large-files/) one could shard across multiple repos or pay a few cents for S3; or perhaps preferably a scientific data repository.  

I suppose an intriguing part of the Contents API is really the write access? The ability to fork, tidy, and push data back is interesting, but sounds like it could also get complicated/messy pretty quickly.  Would have to think a bit more about use cases and workflow for this.  

The csv-to-api thing looks pretty neat; awesome that you've already got working examples of this.  I guess that would work regardless of where the files were hosted?  It would be nice to illustrate the advantage of this though; maybe in reducing unnecessary data transfer on slow connections.

**Comment by sckott** [comment]
@cboettig  Hmmm, so how would you get to files if not through contents API?  Just using raw links like `https://raw.githubusercontent.com/...`?  If so, I was trying that, but it seemed like there were some issues with that approach. It would work 5 times in a row, then fail, then work again, then fail. Anyway, it seemed pretty fragile, but maybe I just screwed something up.  Maybe there's another way you're talking about?

Exploring using s3 definitely seems worth it. 

Right, I gravitated towards Conents API b/c a user could not just read, but write data.  

Yes, the csv-to-api thing should work wherever files are hosted, but e.g., Github gives back base64 encoded data in the contents API, so they'd have to be unencoded first before being passed to the csv2api service as csv2api service expects a csv. I think ideally we'd want our own csv2api service so we aren't potentially crashing someone else's 

p.s. put it in my account, cause just playing here, and not sure if we should all now use onboarding or not...

**Comment by cboettig** [comment]
@sckott  I was thinking of just stick the data on the `gh-pages` branch; the CDN-backed servers behind `gh-pages` seem to be super fast and reliably as far as I can tell.  e.g.: I just put the csv files for rgpdd here: https://github.com/ropensci/rgpdd/tree/gh-pages

so you could do http://ropensci.github.io/rgpdd/data/data.csv etc to pull down the csv files.  Haven't tested it much but I think it should be less fragile than the githubusercontent links.  

Since this hosts the csvs un-encoded I guess it would work more smoothly with the csv2api service? But hosting out own probably makes sense.  

I suppose one might also consider hosting compressed versions of the static csv files (or maybe it makes sense to host them as compressed .rda files if a user really is just pulling them into R ... that data.csv example from GPDD is 330K compressed, so about 1.5% the size of that 21 MB csv file...)

Good point about onboarding; that sounds like a reasonable strategy to me.  How's the actual transfer step work?  Are we asking a user to use the "transfer ownership" mechanism on Github or just forking it once a package is accepted? The trouble with forks is that stars don't transfer over, so it might be bad stats-wise.  (Also the issue-tracker doesn't transfer over).  Anyway, maybe a topic to discuss over in a different issue but thanks for the heads up.

**Comment by sckott** [comment]
@cboettig ah, makes sense now about gh-pages given the CDN backing.   I'll test it out to see how reliable it is, should be good. 

Hmm, `.rda` wouldn't allow anyone from another language to use the files easily though. Seems better to steer away from lang specific files. Anyway, seems that github uses compression by default, and httr does as well, so that helps some

---

Don't think we've talked about it yet, but I think we will ask users to transfer ownership

**Comment by cboettig** [comment]
Sounds good.  Yup, I think we'd always want to provide raw csvs; though if
necessary, we could provide .rdas as well for users concerned about
download speed/bandwidth.

It might be good to have some machine-readable metadata beyond what we can
get with column names, whether EML or something more minimal, at least
identifying the source and citation information for the data.  Of course
we're kinda re-inventing the data repository here...

**Comment by sckott** [comment]
> Of course we're kinda re-inventing the data repository here...

Right, so can we still imagine use cases where people would get use out of this?  Perhaps there is a scenario in which a lab or a person/group of people have data they want to serve up to others, or just make easier to use for themselves, especially if they need flexibility

There is the potential with this to make community contributions much easier with files hosted on github vs. hosted on S3 or other services (including data repos) b/c of the easy to use fork and PR workflow - so I think `living data` repos might have a strong case here

**Comment by cboettig** [comment]
yeah, I really like the living data thing, and the kind of light-weight
flexibility it gives.  I suppose another downside of the GH-pages approach
instead of the github content API is that it really just serves the latest
version, while being able to refer to explicit versions by hash or use a
link that always points to the latest version is a really cool feature
about Github.   Might be nice to develop some use case narratives a bit
more fully.

Maybe we're also in dat territory too here.

**Comment by sckott** [comment]
Good point about `dat`.  I guess `dat` is still not in beta yet, but supposedly should be soon.  I wonder if `dat` will already do what we're talking about here...but i don't know enough. Will hosting `dat` still require a server?  It looks like for very small data can get by with free heroku instance https://github.com/maxogden/dat/blob/master/docs/getting-started.md , but would need e.g. a Droplet for bigger data.  I guess if there's some public dat server that would work, not sure if that exists, or will exist.

Agree that being able to refer to versions is a huge advantage of the content API.
